### DISCUSSION TITLE
Share cached objects between different tty stat

### DISCUSSION THREAD

**Discussion by MizuhaHimuraki** [ccache/ccache#571] [Support]
### How to reproduce ###
a simple c file such as follow

`int main() {
  return 1;
}`

1. ccache clang -c test.c
2. unset TERM
3. ccache clang -c test.c
4. ccache clang -fcolor-diagnostics -c test.c

### Actual behavior ###
3 - cache miss
4 - cache hit(direct)

### Expected behavior ###
3 - cache hit(direct)
4 - cache miss

I found the feature "Automatically enabling colors" may affect the hit rate.
The extra args "-fcolor-diagnostics" added by ccache will be part of the object hash.

details see
https://github.com/ccache/ccache/blob/0b1a0d78bfaac9321fc077464002bd42e6f15e5e/src/ccache.c#L3547-3549

Maybe we should calculate the file hash BEFORE add the option?

**Comment by MizuhaHimuraki** [comment]
I found the solution in Caveats from manual.

Could we have the following solutions?

### solution 1:
add option "sloppiness colors", and then ccache will ignore all color options in object hash

### solution 2: 
always add  "-fcolor-diagnostics=always" and filiter the color format by ccache when the tty is off.

**Comment by jrosdahl** [comment]
> Actual behavior
> 3 - cache miss
> [...]

Like you already found, this is deliberate. I don't consider this to be a bug but a missing feature. See #223 for a proposed solution and #222 for related discussion.

> Expected behavior
> [...]
> 4 - cache miss

Why do you expect a cache miss here?

> solution 1:
> add option "sloppiness colors", and then ccache will ignore all color options in object hash

I would prefer solving the problem for real instead of adding yet another option.

> solution 2:
> always add "-fcolor-diagnostics=always" and filiter the color format by ccache when the tty is off.

Sounds like #223.
